src/Tools/dotnet-monitor/ParameterCapturing/CapturedParametersWriter.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.Options;
+using Microsoft.Diagnostics.Monitoring.WebApi.ParameterCapturing;
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.ParameterCapturing
+{
+    internal sealed class CapturedParametersWriter : IAsyncDisposable
+    {
+        private readonly Channel<ICapturedParameters> _parameters = Channel.CreateUnbounded<ICapturedParameters>();
+        private readonly CancellationToken _cancellationToken;
+        private readonly Task _writerTask;
+        private readonly CapturedParametersFormatter _formatter;
+
+        public CapturedParametersWriter(Stream outputStream, CapturedParameterFormat format, CancellationToken token)
+        {
+            _cancellationToken = token;
+            _formatter = format switch
+            {
+                CapturedParameterFormat.PlainText => new CapturedParametersTextFormatter(outputStream),
+                CapturedParameterFormat.JsonSequence => new CapturedParametersJsonFormatter(outputStream, format),
+                CapturedParameterFormat.NewlineDelimitedJson => new CapturedParametersJsonFormatter(outputStream, format),
+                _ => throw new ArgumentException(nameof(format)),
+            };
+            _writerTask = WriterLoop();
+        }
+
+        public void AddCapturedParameters(ICapturedParameters capturedParameters)
+        {
+            _parameters.Writer.TryWrite(capturedParameters);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            _parameters.Writer.Complete();
+            await _writerTask;
+            await _formatter.DisposeAsync();
+        }
+
+        private async Task WriterLoop()
+        {
+            await _formatter.StartAsync(_cancellationToken);
+
+            while (await _parameters.Reader.WaitToReadAsync(_cancellationToken))
+            {
+                if (_parameters.Reader.TryRead(out ICapturedParameters parameter))
+                {
+                    await _formatter.WriteParameters(parameter, _cancellationToken);
```

Comment:
It seems that if an exception were to be thrown here we'd never propagate up that exception nor recover while continuing to buffer up parameters to be egressed.

I'd recommend we expose the writer task outside of this class and make the `CaptureParametersOption` wait for either the `_capturingStoppedCompletionSource` OR this writer task to complete and propagate up any errors. That way if something goes wrong we report the error and stop the operation.

Reply:
That's a good catch! Please let me know if my implementation is what you had in mind.